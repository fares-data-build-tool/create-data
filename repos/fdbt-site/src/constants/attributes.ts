// Session Attributes

export const FARE_TYPE_ATTRIBUTE = 'fdbt-fare-type';

export const CARNET_FARE_TYPE_ATTRIBUTE = 'fdbt-carnet-fare-type';

export const PASSENGER_TYPE_ATTRIBUTE = 'fdbt-passenger-type';

export const DEFINE_PASSENGER_TYPE_ERRORS_ATTRIBUTE = 'fdbt-passenger-type-errors';

export const SERVICE_ATTRIBUTE = 'fdbt-service';

export const JOURNEY_ATTRIBUTE = 'fdbt-journey';

export const FARE_STAGES_ATTRIBUTE = 'fdbt-fare-stages';

export const STAGE_NAMES_ATTRIBUTE = 'fdbt-stage-names';

export const DURATION_VALID_ATTRIBUTE = 'fdbt-duration-valid';

export const PRODUCT_DETAILS_ATTRIBUTE = 'fdbt-product-details';

export const TICKET_REPRESENTATION_ATTRIBUTE = 'fdbt-ticket-representation';

export const CSV_UPLOAD_ATTRIBUTE = 'fdbt-csv-upload';

export const FARE_ZONE_ATTRIBUTE = 'fdbt-csv-zone-upload';

export const PERIOD_EXPIRY_ATTRIBUTE = 'fdbt-period-expiry';

export const SERVICE_LIST_ATTRIBUTE = 'fdbt-services';

export const INPUT_METHOD_ATTRIBUTE = 'fdbt-input-method';

export const NUMBER_OF_STAGES_ATTRIBUTE = 'fdbt-number-stages';

export const MATCHING_ATTRIBUTE = 'fdbt-matching';

export const INBOUND_MATCHING_ATTRIBUTE = 'fdbt-inbound-matching';

export const NUMBER_OF_PRODUCTS_ATTRIBUTE = 'fdbt-number-of-products';

export const MULTIPLE_PRODUCT_ATTRIBUTE = 'fdbt-multiple-product';

export const PRICE_ENTRY_ATTRIBUTE = 'fdbt-price-entry';

export const SOP_ATTRIBUTE = 'fdbt-sales-offer-package';

export const SOP_INFO_ATTRIBUTE = 'fdbt-sales-offer-package-info';

export const SALES_OFFER_PACKAGES_ATTRIBUTE = 'fdbt-select-sales-offer-packages';

export const GROUP_SIZE_ATTRIBUTE = 'fdbt-group-size';

export const GROUP_PASSENGER_TYPES_ATTRIBUTE = 'fdbt-group-passenger-types';

export const GROUP_PASSENGER_INFO_ATTRIBUTE = 'fdbt-group-passenger-info';

export const GROUP_DEFINITION_ATTRIBUTE = 'fdbt-group-definition';

export const TIME_RESTRICTIONS_DEFINITION_ATTRIBUTE = 'fdbt-time-restrictions-definition';

export const RETURN_VALIDITY_ATTRIBUTE = 'fdbt-return-validity';

export const PRODUCT_DATE_ATTRIBUTE = 'fdbt-product-date-attribute';

export const MULTIPLE_OPERATOR_ATTRIBUTE = 'fdbt-multiple-operator-attribute';

export const MULTIPLE_OPERATORS_SERVICES_ATTRIBUTE = 'fdbt-multiple-operators-services-attribute';

export const FULL_TIME_RESTRICTIONS_ATTRIBUTE = 'fdbt-full-time-restrictions-attribute';

export const TERM_TIME_ATTRIBUTE = 'fdbt-term-time';

export const SCHOOL_FARE_TYPE_ATTRIBUTE = 'fdbt-school-fare-type';

export const FORGOT_PASSWORD_ATTRIBUTE = 'fdbt-reset-password';

export const USER_ATTRIBUTE = 'fdbt-user';

export const OPERATOR_ATTRIBUTE = 'fdbt-operator';

export const TXC_SOURCE_ATTRIBUTE = 'fdbt-txc-source';

export const MULTI_OP_TXC_SOURCE_ATTRIBUTE = 'fdbt-multi-op-txc-source';

export const REUSE_OPERATOR_GROUP_ATTRIBUTE = 'fdbt-reuse-operator-group';

export const SAVE_OPERATOR_GROUP_ATTRIBUTE = 'fdbt-save-operator-group';

export const SAVED_PASSENGER_GROUPS_ATTRIBUTE = 'fdbt-saved-passenger-groups';
