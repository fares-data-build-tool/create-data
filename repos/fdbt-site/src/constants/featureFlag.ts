// const dev = 'dev' === process.env.STAGE;
const devOrTest = ['test', 'dev'].includes(process.env.STAGE ?? '');
console.log(devOrTest);

export const globalSettingsDeleteEnabled = devOrTest;
export const myFaresEnabled = devOrTest;
export const saveProductsEnabled = true;
export const exportEnabled = false;
export const masterStopListEnabled = false;
