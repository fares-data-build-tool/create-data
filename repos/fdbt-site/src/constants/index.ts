export const OPERATOR_COOKIE: string = 'fdbt-operator-cookie';

export const SERVICE_COOKIE: string = 'fdbt-service-cookie';

export const FARETYPE_COOKIE: string = 'fdbt-faretype-cookie';

export const FEEDBACK_LINK: string = 'mailto:fdbt@transportforthenorth.com';

export const GOVUK_LINK: string = 'https://www.nationalarchives.gov.uk/doc/open-government-licence/version/3/';

export const CSV_INPUT_METHOD_COOKIE: string = 'fdbt-csv-input-method-cookie';

export const MANUAL_INPUT_METHOD_COOKIE: string = 'fdbt-manual-input-method-cookie';

export const INT_MAP_INPUT_METHOD_COOKIE: string = 'fdbt-interactive-map-input-method-cookie';
