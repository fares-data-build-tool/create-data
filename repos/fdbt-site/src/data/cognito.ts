import { CognitoIdentityServiceProvider } from 'aws-sdk';
import awsParamStore from 'aws-param-store';
import crypto from 'crypto';

const clientId = process.env.FDBT_USER_POOL_CLIENT_ID as string;
const userPoolId = process.env.FDBT_USER_POOL_ID as string;
let clientSecret: string | undefined;

const getCognitoClient = (): CognitoIdentityServiceProvider =>
    new CognitoIdentityServiceProvider({ region: 'eu-west-2' });

const cognito = getCognitoClient();

const calculateSecretHash = (username: string): string => {
    if (!clientSecret) {
        clientSecret = awsParamStore.getParameterSync('fdbt-cognito-client-secret', { region: 'eu-west-2' }).Value;
    }

    return crypto
        .createHmac('SHA256', clientSecret as string)
        .update(username + clientId)
        .digest('base64');
};

export const initiateAuth = async (
    username: string,
    password: string,
): Promise<CognitoIdentityServiceProvider.AdminInitiateAuthResponse> => {
    console.info('initiating auth...');

    const params: CognitoIdentityServiceProvider.AdminInitiateAuthRequest = {
        AuthFlow: 'ADMIN_USER_PASSWORD_AUTH',
        ClientId: clientId,
        UserPoolId: userPoolId,
        AuthParameters: {
            USERNAME: username,
            PASSWORD: password,
            SECRET_HASH: calculateSecretHash(username),
        },
    };

    try {
        const response = await cognito.adminInitiateAuth(params).promise();

        return response;
    } catch (error) {
        throw new Error(`Failed to authenticate user: ${error.stack}`);
    }
};

export const initiateRefreshAuth = async (
    username: string,
    refreshToken: string,
): Promise<CognitoIdentityServiceProvider.AdminInitiateAuthResponse> => {
    console.info('initiating refresh auth...');

    const params: CognitoIdentityServiceProvider.AdminInitiateAuthRequest = {
        AuthFlow: 'REFRESH_TOKEN_AUTH',
        ClientId: clientId,
        UserPoolId: userPoolId,
        AuthParameters: {
            REFRESH_TOKEN: refreshToken,
            SECRET_HASH: calculateSecretHash(username),
        },
    };

    try {
        const response = await cognito.adminInitiateAuth(params).promise();

        return response;
    } catch (error) {
        throw new Error(`Failed to refresh user session: ${error.stack}`);
    }
};

export const respondToNewPasswordChallenge = async (
    username: string,
    password: string,
    session: string,
): Promise<void> => {
    console.info('new password challenge initiated...');

    const params: CognitoIdentityServiceProvider.AdminRespondToAuthChallengeRequest = {
        ChallengeName: 'NEW_PASSWORD_REQUIRED',
        ClientId: clientId,
        UserPoolId: userPoolId,
        ChallengeResponses: {
            USERNAME: username,
            NEW_PASSWORD: password,
            SECRET_HASH: calculateSecretHash(username),
        },
        Session: session,
    };

    try {
        await cognito.adminRespondToAuthChallenge(params).promise();
    } catch (error) {
        throw new Error(`Failed to respond to password challenge: ${error.stack}`);
    }
};

export const globalSignOut = async (username: string): Promise<void> => {
    console.info('performing global sign out...');

    const params: CognitoIdentityServiceProvider.AdminUserGlobalSignOutRequest = {
        Username: username,
        UserPoolId: userPoolId,
    };

    try {
        await cognito.adminUserGlobalSignOut(params).promise();
    } catch (error) {
        throw new Error(`Failed to perform global sign out: ${error.stack}`);
    }
};

export const updateUserAttributes = async (
    username: string,
    attributes: { Name: string; Value: string }[],
): Promise<void> => {
    console.info('updating user attributes...');

    const params: CognitoIdentityServiceProvider.AdminUpdateUserAttributesRequest = {
        UserAttributes: attributes,
        UserPoolId: userPoolId,
        Username: username,
    };

    try {
        await cognito.adminUpdateUserAttributes(params).promise();
    } catch (error) {
        throw new Error(`Failed to update user attributes: ${error.stack}`);
    }
};

export const forgotPassword = async (username: string): Promise<void> => {
    console.info('start forgot password...');

    const params: CognitoIdentityServiceProvider.ForgotPasswordRequest = {
        ClientId: clientId,
        Username: username,
        SecretHash: calculateSecretHash(username),
    };

    try {
        await cognito.forgotPassword(params).promise();
    } catch (error) {
        throw new Error(`Failed to perform forgot password request: ${error.stack}`);
    }
};

export const confirmForgotPassword = async (
    username: string,
    confirmationCode: string,
    password: string,
): Promise<void> => {
    console.info('confirm forgot password...');

    const params: CognitoIdentityServiceProvider.ConfirmForgotPasswordRequest = {
        ClientId: clientId,
        Username: username,
        ConfirmationCode: confirmationCode,
        Password: password,
        SecretHash: calculateSecretHash(username),
    };

    try {
        await cognito.confirmForgotPassword(params).promise();
    } catch (error) {
        if (error?.code === 'ExpiredCodeException') {
            throw new Error('ExpiredCodeException');
        }
        throw new Error(`Failed to confirm forgotten password: ${error.stack}`);
    }
};

export const updateUserPassword = async (newPassword: string, username: string): Promise<void> => {
    console.info('updating user password...');

    const params: CognitoIdentityServiceProvider.AdminSetUserPasswordRequest = {
        Password: newPassword,
        Permanent: true,
        Username: username,
        UserPoolId: userPoolId,
    };

    try {
        await cognito.adminSetUserPassword(params).promise();
    } catch (error) {
        throw new Error(`Failed to update user password: ${error.stack}`);
    }
};
