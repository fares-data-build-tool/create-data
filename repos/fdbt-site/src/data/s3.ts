import AWS from 'aws-sdk';

export interface FareStage {
    stageName: string;
    prices: {
        price: string;
        fareZones: string[];
    }[];
}

export interface UserFareStages {
    fareStages: FareStage[];
}

export interface UserFareZone {
    FareZoneName: string;
    NaptanCodes: string;
    AtcoCodes: string;
}

const getS3Client = (): AWS.S3 => {
    let options = {};

    if (process.env.NODE_ENV === 'development') {
        options = {
            s3ForcePathStyle: true,
            accessKeyId: 'S3RVER',
            secretAccessKey: 'S3RVER',
            endpoint: new AWS.Endpoint('http://localhost:4572'),
        };
    }

    return new AWS.S3(options);
};

const s3 = getS3Client();

export const getUserFareStages = async (uuid: string): Promise<UserFareStages> => {
    if (!process.env.USER_DATA_BUCKET_NAME) {
        throw new Error('Environment variable for validated bucket not set');
    }

    const params = {
        Bucket: process.env.USER_DATA_BUCKET_NAME,
        Key: `${uuid}.json`,
    };
    const response = await s3.getObject(params).promise();
    const dataAsString = response.Body?.toString('utf-8') ?? '';

    return JSON.parse(dataAsString);
};

export const putStringInS3 = async (
    bucketName: string,
    key: string,
    text: string,
    contentType: string,
): Promise<void> => {
    const request: AWS.S3.Types.PutObjectRequest = {
        Bucket: bucketName,
        Key: key,
        Body: Buffer.from(text, 'binary'),
        ContentType: contentType,
    };

    await s3.putObject(request).promise();
};

export const putDataInS3 = async (
    data: UserFareZone[] | UserFareStages | string,
    key: string,
    processed: boolean,
): Promise<void> => {
    let contentType = '';
    let bucketName = '';

    if (!process.env.USER_DATA_BUCKET_NAME || !process.env.RAW_USER_DATA_BUCKET_NAME) {
        throw new Error('Bucket name environment variables not set.');
    }

    if (processed) {
        bucketName = process.env.USER_DATA_BUCKET_NAME;
        contentType = 'application/json; charset=utf-8';
    } else {
        bucketName = process.env.RAW_USER_DATA_BUCKET_NAME;
        contentType = 'text/csv; charset=utf-8';
    }

    await putStringInS3(bucketName, key, JSON.stringify(data), contentType);
};
