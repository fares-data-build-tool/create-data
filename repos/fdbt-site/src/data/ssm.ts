import { GetParameterCommand, SSMClient } from '@aws-sdk/client-ssm';

const ssm = new SSMClient({ region: 'eu-west-2' });

export const getSsmValue = async (name: string, withDecryption?: boolean): Promise<string> => {
    const input = {
        Name: name,
        WithDecryption: withDecryption ? withDecryption : true,
    };
    const value = (await ssm.send(new GetParameterCommand(input))).Parameter?.Value;

    if (!value) {
        throw new Error(`Parameter ${name} was not found`);
    }

    return value;
};
