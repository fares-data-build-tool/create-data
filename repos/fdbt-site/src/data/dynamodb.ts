import AWS from 'aws-sdk';
import dateFormat from 'dateformat';
import flatMap from 'array.prototype.flatmap';
import { NAPTAN_TABLE_NAME, SERVICES_TABLE_NAME, TNDS_TABLE_NAME, NAPTAN_TABLE_GSI } from '../constants';

export interface ServiceType {
    lineName: string;
    startDate: string;
}

export interface JourneyPattern {
    startPoint: {
        Id: string;
        Display: string;
    };
    endPoint: {
        Id: string;
        Display: string;
    };
    stopList: string[];
}

export interface RawJourneyPatternSection {
    Id: string;
    OrderedStopPoints: {
        StopPointRef: string;
        CommonName: string;
    }[];
    StartPoint: string;
    EndPoint: string;
}

export interface RawJourneyPattern {
    JourneyPatternSections: RawJourneyPatternSection[];
}

interface DynamoNaptanInfo {
    CommonName: string;
    NaptanCode: string;
    ATCOCode: string;
    NptgLocalityCode: string;
    LocalityName: string;
    Indicator: string;
    Street: string;
}

interface DynamoNaptanIndex {
    naptanCode: string;
    atcoCode: string;
}

interface RawDynamoNaptanIndex {
    NaptanCode: string;
    Partition: string;
}

export interface Stop {
    stopName: string;
    naptanCode: string;
    atcoCode: string;
    localityCode: string;
    localityName: string;
    indicator: string;
    street: string;
    qualifierName?: string;
}

export interface StopIdentifiers {
    naptanCode: string | null;
    atcoCode: string;
}

export interface Service {
    serviceDescription: string;
    operatorShortName: string;
    journeyPatterns: JourneyPattern[];
}

export interface RawService {
    serviceDescription: string;
    operatorShortName: string;
    journeyPatterns: RawJourneyPattern[];
}

const getDynamoDBClient = (): AWS.DynamoDB.DocumentClient => {
    const dynamoDbRegion = process.env.AWS_REGION || 'eu-west-2';

    const options = {
        convertEmptyValues: true,
        region: dynamoDbRegion,
    };

    let client = null;

    if (process.env.NODE_ENV === 'development') {
        client = new AWS.DynamoDB.DocumentClient({
            ...options,
            endpoint: 'http://localhost:9100',
        });
    } else {
        client = new AWS.DynamoDB.DocumentClient(options);
    }

    return client;
};

const dynamoDbClient = getDynamoDBClient();

export const convertDateFormat = (startDate: string): string => {
    return dateFormat(startDate, 'dd/mm/yyyy');
};

export const getServicesByNocCode = async (nocCode: string): Promise<ServiceType[]> => {
    const tableName = process.env.NODE_ENV === 'development' ? 'dev-Services' : SERVICES_TABLE_NAME;

    const queryInput: AWS.DynamoDB.DocumentClient.QueryInput = {
        TableName: tableName,
        KeyConditionExpression: '#pk = :value',
        ExpressionAttributeNames: {
            '#pk': 'Partition',
        },
        ExpressionAttributeValues: {
            ':value': nocCode,
        },
    };

    try {
        const { Items } = await dynamoDbClient.query(queryInput).promise();
        return (
            Items?.map(
                (item): ServiceType => ({ lineName: item.LineName, startDate: convertDateFormat(item.StartDate) }),
            ) || []
        );
    } catch (err) {
        throw new Error(`Could not retrieve services from DynamoDB: ${err.name}, ${err.message}`);
    }
};

export const batchGetStopsByAtcoCode = async (atcoCodes: string[]): Promise<Stop[] | []> => {
    const tableName = process.env.NODE_ENV === 'development' ? 'dev-Stops' : NAPTAN_TABLE_NAME;
    const count = atcoCodes.length;
    const batchSize = 100;
    const batchArray = [];

    for (let i = 0; i < count; i += batchSize) {
        batchArray.push(atcoCodes.slice(i, i + batchSize));
    }

    const batchPromises = batchArray.map(batch => {
        const batchQueryInput: AWS.DynamoDB.DocumentClient.BatchGetItemInput = {
            RequestItems: {
                [tableName]: {
                    ExpressionAttributeNames: {
                        '#in': 'Indicator',
                    },
                    Keys: batch.map(code => ({
                        Partition: code,
                    })),
                    ProjectionExpression: 'LocalityName,#in,Street,CommonName, NaptanCode, ATCOCode, NptgLocalityCode',
                },
            },
        };

        return dynamoDbClient.batchGet(batchQueryInput).promise();
    });

    try {
        const results = await Promise.all(batchPromises);
        const filteredResults = results.filter(item => item.Responses?.[tableName]);
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const naptanItems: DynamoNaptanInfo[] = flatMap(filteredResults, (item: any) => item.Responses[tableName]);

        return naptanItems.map(item => ({
            stopName: item.CommonName,
            naptanCode: item.NaptanCode,
            atcoCode: item.ATCOCode,
            localityCode: item.NptgLocalityCode,
            localityName: item.LocalityName,
            indicator: item.Indicator,
            street: item.Street,
        }));
    } catch (error) {
        console.error(`Error performing batch get for naptan info for stop list: ${atcoCodes}, error: ${error}`);
        throw new Error(error);
    }
};

export const getAtcoCodesByNaptanCodes = async (naptanCodes: string[]): Promise<DynamoNaptanIndex[]> => {
    const tableName = process.env.NODE_ENV === 'development' ? 'dev-Stops' : NAPTAN_TABLE_NAME;
    const indexName = process.env.NODE_ENV === 'development' ? 'NaptanIndex' : NAPTAN_TABLE_GSI;

    const queryPromises = naptanCodes.map(async naptanCode => {
        const queryInput: AWS.DynamoDB.DocumentClient.QueryInput = {
            TableName: tableName,
            IndexName: indexName,
            KeyConditionExpression: 'NaptanCode = :v_code',
            ExpressionAttributeValues: {
                ':v_code': naptanCode,
            },
        };
        return dynamoDbClient.query(queryInput).promise();
    });
    try {
        const results = await Promise.all(queryPromises);
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const atcoItems: RawDynamoNaptanIndex[] = flatMap(results, (item: any) => item.Items);
        return atcoItems.map(item => ({ atcoCode: item.Partition, naptanCode: item.NaptanCode }));
    } catch (error) {
        console.error(
            `Error performing queries for ATCO Codes using Naptan Codes: ${naptanCodes}. Error: ${error.stack}`,
        );
        throw new Error(error);
    }
};

export const getServiceByNocCodeAndLineName = async (nocCode: string, lineName: string): Promise<RawService> => {
    const tableName = process.env.NODE_ENV === 'development' ? 'dev-TNDS' : TNDS_TABLE_NAME;

    const queryInput: AWS.DynamoDB.DocumentClient.QueryInput = {
        TableName: tableName,
        KeyConditionExpression: '#pkAttNm = :pkAttVal and begins_with (#skAttNm, :skAttVal)',
        ExpressionAttributeNames: {
            '#pkAttNm': 'Partition',
            '#skAttNm': 'Sort',
        },
        ExpressionAttributeValues: {
            ':pkAttVal': nocCode,
            ':skAttVal': lineName,
        },
    };
    let Items;

    try {
        ({ Items } = await getDynamoDBClient()
            .query(queryInput)
            .promise());
    } catch (err) {
        throw new Error(`Could not get journey patterns from Dynamo DB: ${err.name}, ${err.message}`);
    }

    const service = Items?.[0];

    if (!service || !service.JourneyPatterns || service.JourneyPatterns.length === 0) {
        throw new Error(`No journey patterns found for nocCode: ${nocCode}, lineName: ${lineName}`);
    }

    return {
        serviceDescription: service.ServiceDescription,
        operatorShortName: service.OperatorShortName,
        journeyPatterns: service.JourneyPatterns,
    };
};
