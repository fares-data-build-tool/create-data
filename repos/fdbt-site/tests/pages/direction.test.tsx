import * as React from 'react';
import { shallow } from 'enzyme';
import { NextPageContext } from 'next';
import { mockRequest } from 'mock-req-res';
import MockRes from 'mock-res';

import Direction from '../../src/pages/direction';
import { OPERATOR_COOKIE, SERVICE_COOKIE } from '../../src/constants';
import { getJourneyPatternsAndLocalityByNocCodeAndLineName, ServiceInformation } from '../../src/data/dynamodb';

jest.mock('../../src/data/dynamodb');

const mockServiceInfo: ServiceInformation = {
    serviceDescription: 'Harrogate - Beckwith Knowle',
    journeyPatterns: [
        {
            startPoint: {
                Display: 'Harrogate Bus Station Stand 5, Harrogate',
                Id: '3200YND90060',
            },
            endPoint: {
                Display: 'Harrogate Bus Station Stand 6, Harrogate',
                Id: '3200YND90060',
            },
        },
        {
            startPoint: {
                Display: 'Harrogate Place 1, Harrogate',
                Id: '000HGTPL1000',
            },
            endPoint: {
                Display: 'Harrogate Place 2, Harrogate',
                Id: '000HGTPL2000',
            },
        },
        {
            startPoint: {
                Display: 'Pannal, Harrogate',
                Id: '000PNAL0010',
            },
            endPoint: {
                Display: 'Knaresborough Bus Station, Harrogate',
                Id: '000KSBRG0020',
            },
        },
    ],
};

describe('pages', () => {
    describe('direction', () => {
        beforeEach(() => {
            (({ ...getJourneyPatternsAndLocalityByNocCodeAndLineName } as jest.Mock).mockImplementation(
                () => mockServiceInfo,
            ));
        });

        it('should render correctly', () => {
            const tree = shallow(
                <Direction Operator="Connexions Buses" lineName="X6A" serviceInfo={mockServiceInfo} />,
            );
            expect(tree).toMatchSnapshot();
        });

        it('shows operator name above the select box', () => {
            const wrapper = shallow(
                <Direction Operator="Connexions Buses" lineName="X6A" serviceInfo={mockServiceInfo} />,
            );
            const journeyWelcome = wrapper.find('#direction-operator-linename-hint').first();

            expect(journeyWelcome.text()).toBe('Connexions Buses - X6A');
        });
        it('shows a list of journey patterns for the service in the select box', () => {
            const wrapper = shallow(
                <Direction Operator="Connexions Buses" lineName="X6A" serviceInfo={mockServiceInfo} />,
            );
            const serviceJourney = wrapper.find('.journey-option');

            expect(serviceJourney).toHaveLength(3);
            expect(serviceJourney.first().text()).toBe(
                'Harrogate Bus Station Stand 5, Harrogate TO Harrogate Bus Station Stand 6, Harrogate',
            );
            expect(serviceJourney.at(1).text()).toBe('Harrogate Place 1, Harrogate TO Harrogate Place 2, Harrogate');
            expect(serviceJourney.at(2).text()).toBe('Pannal, Harrogate TO Knaresborough Bus Station, Harrogate');
        });

        it('returns operator value and list of services when operator cookie exists with NOCCode', async () => {
            const operator = 'HCTY';
            const lineName = 'X6A';

            const res = new MockRes();

            const req = mockRequest({
                connection: {
                    encrypted: false,
                },
                headers: {
                    host: 'localhost:5000',
                    cookie: `${OPERATOR_COOKIE}=%7B%22operator%22%3A%22${operator}%22%2C%22uuid%22%3A%221e0459b3-082e-4e70-89db-96e8ae173e10%22%2C%22nocCode%22%3A%22HCTY%22%7D; fdbt-faretype-cookie=%7B%22faretype%22%3A%22single%22%2C%22uuid%22%3A%221e0459b3-082e-4e70-89db-96e8ae173e10%22%7D; ${SERVICE_COOKIE}=%7B%22service%22%3A%22${lineName}%2329%2F04%2F2019%22%2C%22uuid%22%3A%221e0459b3-082e-4e70-89db-96e8ae173e10%22%7D`,
                },
                cookies: {
                    OPERATOR_COOKIE: operator,
                    SERVICE_COOKIE: lineName,
                },
            });
            const ctx: NextPageContext = {
                res,
                req,
                pathname: '',
                query: {},
                AppTree: () => <div />,
            };
            const result = await Direction.getInitialProps(ctx);
            expect(result).toEqual({
                Operator: operator,
                lineName,
                serviceInfo: mockServiceInfo,
            });
        });
        it('redirects to the error page if no journey patterns can be found', async () => {
            (({ ...getJourneyPatternsAndLocalityByNocCodeAndLineName } as jest.Mock).mockImplementation(() =>
                Promise.resolve(null),
            ));
            const operator = 'HCTY';
            const lineName = 'X6A';

            const mockWriteHeadFn = jest.fn();
            const mockEndFn = jest.fn();
            const res = new MockRes();
            res.writeHead = mockWriteHeadFn;
            res.end = mockEndFn;

            const req = mockRequest({
                connection: {
                    encrypted: false,
                },
                headers: {
                    host: 'localhost:5000',
                    cookie: `${OPERATOR_COOKIE}=%7B%22operator%22%3A%22${operator}%22%2C%22uuid%22%3A%221e0459b3-082e-4e70-89db-96e8ae173e10%22%2C%22nocCode%22%3A%22HCTY%22%7D; fdbt-faretype-cookie=%7B%22faretype%22%3A%22single%22%2C%22uuid%22%3A%221e0459b3-082e-4e70-89db-96e8ae173e10%22%7D; ${SERVICE_COOKIE}=%7B%22service%22%3A%22${lineName}%2329%2F04%2F2019%22%2C%22uuid%22%3A%221e0459b3-082e-4e70-89db-96e8ae173e10%22%7D`,
                },
                cookies: {
                    OPERATOR_COOKIE: operator,
                    SERVICE_COOKIE: lineName,
                },
            });

            const ctx: NextPageContext = {
                res,
                req,
                pathname: '',
                query: {},
                AppTree: () => <div />,
            };
            const result = await Direction.getInitialProps(ctx);

            expect(mockWriteHeadFn).toHaveBeenCalledWith(302, {
                Location: '/error',
            });
            expect(mockEndFn).toHaveBeenCalled();
            expect(result).toEqual({});
        });
        it('redirects to the error page if the operator or service cookies do not exist', async () => {
            const operator = 'HCTY';
            const lineName = 'X6A';

            const mockWriteHeadFn = jest.fn();
            const mockEndFn = jest.fn();
            const res = new MockRes();
            res.writeHead = mockWriteHeadFn;
            res.end = mockEndFn;

            const req = mockRequest({
                connection: {
                    encrypted: false,
                },
                headers: {
                    host: 'localhost:5000',
                    cookie: `cookieOne=%7B%22operator%22%3A%22${operator}%22%2C%22uuid%22%3A%221e0459b3-082e-4e70-89db-96e8ae173e10%22%2C%22nocCode%22%3A%22HCTY%22%7D; fdbt-faretype-cookie=%7B%22faretype%22%3A%22single%22%2C%22uuid%22%3A%221e0459b3-082e-4e70-89db-96e8ae173e10%22%7D; cookieTwo=%7B%22service%22%3A%22${lineName}%2329%2F04%2F2019%22%2C%22uuid%22%3A%221e0459b3-082e-4e70-89db-96e8ae173e10%22%7D`,
                },
                cookies: {
                    OPERATOR_COOKIE: operator,
                    SERVICE_COOKIE: lineName,
                },
            });

            const ctx: NextPageContext = {
                res,
                req,
                pathname: '',
                query: {},
                AppTree: () => <div />,
            };
            const result = await Direction.getInitialProps(ctx);

            expect(mockWriteHeadFn).toHaveBeenCalledWith(302, {
                Location: '/error',
            });
            expect(mockEndFn).toHaveBeenCalled();
            expect(result).toEqual({});
        });
    });
});
