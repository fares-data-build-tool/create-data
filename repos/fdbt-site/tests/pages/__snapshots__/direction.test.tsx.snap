// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages direction should render correctly 1`] = `
<Layout
  description="Direction selection page of the Fares data build tool"
  title="Select a Direction - Fares data build tool"
>
  <main
    className="govuk-main-wrapper app-main-class"
    id="main-content"
    role="main"
  >
    <form
      action="/api/direction"
      method="post"
    >
      <div
        className="govuk-form-group"
      >
        <fieldset
          aria-describedby="page-heading"
          className="govuk-fieldset"
        >
          <legend
            className="govuk-fieldset__legend govuk-fieldset__legend--xl"
          >
            <h1
              className="govuk-fieldset__heading"
              id="page-heading"
            >
              Please select your journey direction
            </h1>
          </legend>
          <span
            className="govuk-hint"
            id="direction-operator-linename-hint"
          >
            Connexions Buses
             - 
            X6A
          </span>
          <span
            className="govuk-hint"
            id="direction-journey-description-hint"
          >
            Journey: Harrogate - Beckwith Knowle
          </span>
          <select
            className="govuk-select"
            defaultValue=""
            id="journeyPattern"
            name="journeyPattern"
          >
            <option
              disabled={true}
              value=""
            >
              Select One
            </option>
            <option
              className="journey-option"
              key="Harrogate Bus Station Stand 5, Harrogate"
              value="Harrogate Bus Station Stand 6, Harrogate"
            >
              Harrogate Bus Station Stand 5, Harrogate
               TO 
              Harrogate Bus Station Stand 6, Harrogate
            </option>
          </select>
        </fieldset>
      </div>
      <input
        className="govuk-button govuk-button--start"
        id="continue-button"
        type="submit"
        value="Continue"
      />
    </form>
  </main>
</Layout>
`;
